### ISSUE TITLE
Mex interface solution inconsistency

### ISSUE THREAD

**Issue by msvec94** [oxfordcontrol/osqp#205]
When I was testing parameter update, I found out that solutions of my problems vary, although the problem remained basically the same during consecutive runs. What I did was:

1. Created OSQP problem with predefined matrices and solved it (`resultsOne.x`).
2. Generated mex interface from that problem and solved this new embedded problem (`xOne`).
3. Updated original problem parameters using matrices and vectors which weren't changed to test the functionality and solved "modified" problem (`resultsTwo.x`).
4. Updated generated mex interface problem using the same methodology as in step 3. and solved that one as well (`xTwo`).

After comparing solutions to each other, I found out that only `resultsOne.x` and `resultsTwo.x` were (almost) the same (I assume the differences come from the fact that warm_start flag is on, so the second solution is obtained sooner and represents more accurate approximation of the theoretic solution). However, `xOne` is simmilar neither to `xTwo`, nor to `resultsOne.x` and `resultsTwo.x` . All in all, the results differ when they should be exactly the same.

I have prepared some example code to demonstrate mentioned behavior. Error vectors `errOne`, `errTwo`, `crossErr` and `crossErrEm` should all be zero, but they are not.  I have attached [data.zip](https://github.com/oxfordcontrol/osqp/files/3616098/data.zip) file (which contains data.mat) so you can load test data.

_CODE:_
`load('data.mat'); %Load test data`

`prob = osqp;`
`% Matrix P=Px+Px' to ensure it is symmetric`
`prob.setup(sparse(Px+Px'), q, sparse(Ax), low, upp, 'eps_abs', 1e-07, 'eps_rel', 1e-07, 'max_iter', 5000);`

`% Used to generate embedded problem`
`%prob.codegen('code', 'parameters', 'matrices', 'mexname', 'emosqp');`

`% Solve original problem`
`resultsOne = prob.solve();`
`[xOne, yOne, statusValOne, iterOne, runTimeOne] = emosqp('solve');`

`% "Update matrices" without chaning the values`
`Px_new = Px+Px';`
`Ax_new = Ax;`
`q_new = q;`
`low_new = low;`

`prob.update('q', q_new);`
`prob.update('Px', nonzeros(triu(Px_new)));`
`prob.update('l', low_new);`
`prob.update('Ax', nonzeros(Ax_new));`

`emosqp('update_lin_cost', q_new);`
`emosqp('update_lower_bound', low_new);`
`emosqp('update_P', nonzeros(triu(Px_new)), [], 0);`
`emosqp('update_A', nonzeros(Ax_new), [], 0);`

`% Solve updated problem`
`resultsTwo = prob.solve();`
`[xTwo, yTwo, statusValTwo, iterTwo, runTimeTwo] = emosqp('solve');`

`% Error check`
`errOne = abs(resultsOne.x - xOne);`
`errTwo = abs(resultsTwo.x - xTwo);`
`crossErr = abs(resultsOne.x - resultsTwo.x);`
`crossErrEm = abs(xOne - xTwo);`

**Comment by goulart-paul** [comment]
Your problem appears to be either infeasible or unbounded, since OSQP never converges even after a huge number of iterations, but is never actually declared unsolvable by the convergence checks.   Perhaps the issue is tangentially related to #180?   @gbanjac : maybe this is a good test problem for that issue.

**Comment by msvec94** [comment]
I am aware of that fact. However, `prob.solve()` always gives a solution. The first time in a lot of iterations, the second time after 25 iterations, whereas the code generated solver reaches maximum iteration number every time. My question was regarding the fact that code generated solver doesn't behave the same as the original solver, which is used to create it. Why is that so?

**Comment by goulart-paul** [comment]
Perhaps it is related to this?: https://github.com/oxfordcontrol/osqp/issues/203 

For which a fix was provided here: https://github.com/oxfordcontrol/osqp-matlab/pull/18

**Comment by msvec94** [comment]
This is an issue posted by my colleague and it is not a problem anymore.  The problem we have at the moment is a different issue.  For a given test case, `prob.solve()` and `emosqp('solve')` simply don't yield the same solution (`prob.solve()` one actually solves the problem, whereas `emosqp('solve')` does not and reaches maximum number of iterations). This is what I am trying to figure out, since those solvers should be the same and, therefore, have the same solutions).

**Comment by bstellato** [comment]
Could it be related to https://osqp.discourse.group/t/optimization-of-generated-code/39/9 ?

**Comment by goulart-paul** [comment]
I understand, but I'm asking about application of the fix at https://github.com/oxfordcontrol/osqp-matlab/pull/18 because, without applying it, it's possible that the solver is overwriting data in places it should not.   That is one possible explanation of the results you are seeing.

If it is not that, then I suggest you see if you can recreate the problem on a QP of minimal size (i.e. 2-3 variables, 2-3 constraints), then let us have a look at the code that is being exported so we can see if the codegen is breaking somehow.

**Comment by msvec94** [comment]
I have created a small QP where the problem is partially present. The difference is that, in this case, the results are almost similar (error up to 1e-6, which is negligible). However, the code generated solver still reaches maximum number of iterations (in my case 1000), whereas `prob.solve()` solves the problem in 350 iterations. I have included necessary data in [data.zip](https://github.com/oxfordcontrol/osqp/files/3636364/data.zip) file.


_CODE:_

`load('test');`

`prob = osqp;`
`prob.setup(sparse(P), q, sparse(A), l, u, 'eps_abs', 1e-8, 'eps_rel', 1e-8, 'max_iter', 1000);`

`%prob.codegen('code', 'parameters', 'matrices', 'mexname', 'emosqp');`

`results = prob.solve();`
`[x, y, statusVal, iter, runTime] = emosqp('solve');`

`% Error check`
`err = abs(results.x - x);`

**Comment by goulart-paul** [comment]
The problem is that you are compiling in the C code with `EMBEDDED` option defined as `2`, which means that no rho updates / refactorizations are performed.  

The behaviour of the two methods is identical for me if I  replace your matlab call above with 
`prob.setup(sparse(P), q, sparse(A), l, u, 'eps_abs', 1e-8, 'eps_rel', 1e-8, 'max_iter', 5000, 'adaptive_rho', 0 );`

In that case both solvers terminate after 1225 iterations and with identical primal and dual residuals.  Note that I had to also manually increase the maximum iterations within the generated `workspace.c` file, since I did not regenerate your C code.
